File: MGToolbox/MGProgramFiles/MGIntegrator.c
/* -------------------------------------------------------------------------- **
** PURPOSE:  Solve a set of first order ordinary differential equations of    **
**           the form dy(i)/dt = f( t, y(1), ..., y(n) )  (i = 1 ... n)       **
**           with a Runga-Kutta-Merson algorithm that numerically integrates  **
**           the differential equations from tStart to tStart + hEntry.       **
**           (L. Fox, Numerical Solutions of Ordinary and Partial Differ-     **
**           ential Equations, Palo Alto: Addison-Wesley, 1962, pp. 24-25)    **
**                                                                            **
** INPUT:                                                                     **
**     eqns  Function that evaluates dy(i)/dt (i = 1 ... n), the first        **
**           derivatives of y(1) ... y(n) with respect to t.                  **
**                                                                            **
**     numY  The number n of differential equations to be solved.             **
**        y  One-dimensional array whose elements are y(1) ... y(n).          **
**                                                                            **
**   tStart  Value of independent variable t on entry to this function.       **
**                                                                            **
**   hEntry  Suggested numerical integration stepsize for this function.      **
**           If *hEntry == 0, the function eqns is called and dy(i)/dt        **
**           (i = 1 ... n) are evaluated, but no integration is performed.    **
**           On return, *hEntry is the actual stepsize the integrator used.   **
**                                                                            **
**    hNext  Returns estimated stepsize for the next call to this function.   **
**                                                                            **
** absError  Allowable absolute error in y(i)  (i = 1 ... n).                 **
** relError  Allowable relative error in y(i)  (i = 1 ... n).                 **
**                                                                            **
** OUTPUT:   Returns an error message if integration failed, otherwise NULL.  **
**       y   The values of y(i) (i=1 ... n) at t + *hEntry are returned in y. **
**  hEntry   The actual stepsize used by integrator is returned in hEntry.    **
**   hNext   The integrator's estimate (based on error analysis of current    **
**           step) is returned in hNext, or 0 if integration failed.          **
**                                                                            **
** Copyright (c) 2009-2021 by Paul Mitiguy.  Use permitted under the 3-clause **
**           BSD license: https://opensource.org/licenses/BSD-3-Clause.       **
**           This copyright notice must appear in all copies & distributions. **
**           This copyright notice applies to the functions MGIntegrator,     **
**           MGIntegrateOneStep, MGIntegrateForwardOrBackward                 **
** -------------------------------------------------------------------------- */
static char*  MGIntegrator( char*(*eqns)(double, double*, double*, char),
                            int numY, double y[], double tStart, double* hEntry, double* hNext,
                            double smallestAllowableStepsize, double absError, double relError )
{
   double f0[myNumberOfODES], f1[myNumberOfODES], f2[myNumberOfODES];
   double y1[myNumberOfODES], y2[myNumberOfODES];
   double h = hEntry ? *hEntry : 0;

   /* Always calculate derivatives at tStart (integration boundary here).   */
   /* If h == 0, just call eqns at tStart and return.                       */
   char* errorMessage = (*eqns)( tStart, y, f0, 1 );
   if( !errorMessage  &&  h == 0 )  return NULL;

   /* Avoid endless loop due to adding tiny h to tStart (precision problem) */
   while( !errorMessage  &&  tStart + 0.1*h != tStart )
   {
      int    i, indexFailedVariable = -1;        /* Variable that failed.   */
      double errorRatioMax = 0;                  /* Testing error criterion */
      double h2 = h * 0.5;                                 /* Half    of h  */
      double h3 = h / 3.0;                                 /* Third   of h  */
      double h6 = h / 6.0;                                 /* Sixth   of h  */
      double h8 = h * 0.125;                               /* Eighth  of h  */
      for( i=0;  i<numY;  i++ )  y1[i] = y[i] + h3*f0[i];
      if( (errorMessage = (*eqns)( tStart+h3, y1, f1, 0 )) != NULL ) break;
      for( i=0;  i<numY;  i++ )  y1[i] = y[i] + h6*(f0[i] + f1[i]);
      if( (errorMessage = (*eqns)( tStart+h3, y1, f1, 0 )) != NULL ) break;
      for( i=0;  i<numY;  i++ )  y1[i] = y[i] + h8*(f0[i] + 3*f1[i]);
      if( (errorMessage = (*eqns)( tStart+h2, y1, f2, 0 )) != NULL ) break;
      for( i=0;  i<numY;  i++ )  y1[i] = y[i] + h2*(f0[i] - 3*f1[i] + 4*f2[i]);
      if( (errorMessage = (*eqns)( tStart+h,  y1, f1, 0 )) != NULL)  break;
      for( i=0;  i<numY;  i++ )  y2[i] = y[i] + h6*(f0[i] + 4*f2[i] + f1[i]);

      /* Both y1[i] and y2[i] provide estimates to the new value of y.      */
      /* Decide if the relative and absolute error tolerances are met.      */
      /* If they are not, reduce the stepsize and restart the integration.  */
      for( i = 0;  i < numY;  i++ )              /* Check all variables     */
      {
         double errorInEstimate = 0.2 * fabs( y2[i] - y1[i] );
         double relTest  = fabs( y2[i] ) * relError;
         double largerOfAbsErrOrRelTest = absError > relTest ? absError : relTest;
         double errorRatio = errorInEstimate / largerOfAbsErrOrRelTest;
         if( errorRatio > errorRatioMax ) { errorRatioMax = errorRatio;  indexFailedVariable = i; }
      }

      /* If integration succeeded, update values of y and t before return.  */
      /* Return actual stepsize and estimate for next integration stepsize. */
      if( errorRatioMax < 1 )
      {
         for( i = 0;  i < numY;  i++ )  y[i] = y2[i];
         *hEntry = h;   *hNext = (errorRatioMax < 1.0/64.0) ? 2*h : h;
         return NULL;
      }

      /* Otherwise, errorRatioMax >= 1, so absError or relTest failed.      */
      /* In other words, errorInEstimate >= largerOfAbsErrOrRelTest.        */
      /* Try to halve stepsize and restart integration.                     */
      if( fabs(h=h2) <= fabs(smallestAllowableStepsize) )
      {
         static char stepsizeCutMessage[128];
         sprintf(errorMessage = stepsizeCutMessage,
                 "Error: Numerical integration stepsize cut too many times at t = %17.9E (variable %d).",
                 tStart, indexFailedVariable);
      }
   }

   /* Check if loop terminated due to numerical round-off.                  */
   if( !errorMessage )
      errorMessage = "Error: Numerical round off makes stepsize h too small relative to tStart, so tStart+h = tStart."
                   "\nIntegration stepsize may have been cut too many times.";

   /* Print error message that numerical integrator failed.    */
   /* If h != 0, call eqns to fill for error display.          */
   printf( "\n Error: Numerical integration failed at t = %17.9E.\n\n", tStart );
   if( h != 0 ) (*eqns)( tStart, y, f0, 1 );
   return errorMessage;
}


/* -------------------------------------------------------------------------- **
** PURPOSE:  Controls the independent variable, e.g., t while solving a set   **
**           of first order ordinary differential equations of the form       **
**           dy(i)/dt = f( t, y(1), ..., y(n) )  (i = 1, ..., n).             **
**           Discontinuities in functions and t should be handled here.       **
**                                                                            **
** INPUT:    eqns, numY, y, absError, relError are described above.           **
**           If numY = 0, the static value of myPreviousStepsize is reset to  **
**           tStep.  This is useful when integrating several sets of ODEs.    **
**                                                                            **
**        t  The current value of the independent variable.                   **
** tStepMax  Maximum integration stepsize for this integrator step.  On entry **
**           if tStepMax = 0, eqns is called and dy(i)/dt (i=1 ... n) are     **
**           evaluated, but no integration performed.                         **
**                                                                            **
** OUTPUT:   Returns an error message if integration failed, otherwise NULL.  **
**        t  The value of  t + tStep  is returned in t.                       **
**        y  The values of y(i) (i=1 ... n) at t + tStep are returned in y.   **
** -------------------------------------------------------------------------- */
static char*  MGIntegrateOneStep( char*(*eqns)(double, double*, double*, char),
                                  int numY, double y[], double* t, double tStepMax,
                                  double* stepsizeSuggested, double smallestAllowableStepsize,
                                  double absError, double relError )
{
   double hAccumulated = 0;                      /* How far to tStepMax.    */
   double h = *stepsizeSuggested;                /* Current stepsize.       */
   int    isStepFinished = (tStepMax == 0);

   /* Make as many little integration steps as necessary to get to end.     */
   /* Each integration step starts at *t and ends at *t + h.                */
   while( !isStepFinished )
   {
      /* Numerically integrate y[i] and maybe get a smaller value of h.     */
      /* Set hNext to integrator's estimate of next integration step-size.  */
      double hBeforeCall = h, hNext;             /* Suggested stepsize.     */
      char* errorMessage = MGIntegrator( eqns, numY, y, *t, &h, &hNext, smallestAllowableStepsize, absError, relError );
      if( errorMessage ) return errorMessage;    /* Integration failed      */

      /* Any time or function discontinuities should be handled here.       */
      /* Increment value of t using stepsize returned by the integrator.    */
      *t += h;                                   /* Prepare for next step.  */

      /* Change the stepsize if not taking the last step.                   */
      /* Reduce the stepsize if the integrator reduced the stepsize.        */
      /* Increase the stepsize if the integrator suggests a larger one.     */
      /* Update stepsizeSuggested for next call to integrator.              */
      /* Ensure stepsizeSuggested is not larger than tStepMax.              */
      isStepFinished = fabs( (hAccumulated+=h) + smallestAllowableStepsize) > fabs(tStepMax);
      if( !isStepFinished  ||  fabs(h) < fabs(hBeforeCall)  ||  fabs(hNext) > fabs(*stepsizeSuggested) )
      {
         if( fabs(hNext) > fabs(tStepMax) ) hNext = tStepMax;
         *stepsizeSuggested = h = hNext;
      }

      /* Integrator may suggest a stepsize that is bigger than allowed.     */
      /* If next jump is past or very close to end of tStepMax, adjust h.   */
      if( fabs(hAccumulated + 1.1*h) > fabs(tStepMax) )  h = tStepMax - hAccumulated;
   }

   /* Integration finished.  Calculate derivatives of y at final value of t */
   return MGIntegrator( eqns, numY, y, *t, NULL, NULL, 0, 0, 0 );
}


/* ------------------------------------------------------------------------ */
static char*  MGIntegrateForwardOrBackward( int numVariables, double varArrayToIntegrate[], double outputToFill[], double tInitial, double tFinal, double tStepMax, double absError, double relError, int printIntScreen, int printIntFile )
{
   double stepsizeSuggested = tStepMax,   smallestAllowableStepsize = 1.0E-7 * tStepMax;
   double t = tInitial,   tFinalMinusEpsilon = tFinal - smallestAllowableStepsize;
   int    isFirstCall = 1,  isIntegrateForward = (tFinal > tInitial),  integrateDirection = isIntegrateForward ? 1 : -1;
   int    isPrintToScreen,  isPrintToFile,  printCounterScreen = 0,   printCounterFile = 0;

   /* Ensure valid parameters and sign(tStepMax) moves integration in proper direction. */
   char* errorMessage = (numVariables <= 0  ||  !varArrayToIntegrate  ||  absError <= 0  || relError < 0  ||  integrateDirection * tStepMax < 0) ?
                        "Error: Invalid argument to MGIntegrateForwardOrBackward" : NULL;

   /* Initialize integrator with call at t = tInitial, thereafter integrate */
   int isIntegrationFinished = 0;
   while( !isIntegrationFinished  &&  !errorMessage )
   {
      /* Near the end of numerical integration, perhaps take a partial step (decrease tStepMax). */
      if( (isIntegrateForward && t+tStepMax > tFinal)  ||  (!isIntegrateForward && t+tStepMax < tFinal) )
      {
         tStepMax = tFinal - t;
         if( integrateDirection * stepsizeSuggested > fabs(tStepMax) )  stepsizeSuggested = tStepMax;
      }
      /* Due to round-off or other, maybe set stepsizeSuggested = tStepMax. */
      else if( fabs(stepsizeSuggested) >= 0.99 * fabs(tStepMax) ) stepsizeSuggested = tStepMax;

      errorMessage = MGIntegrateOneStep( MGeqns, numVariables, varArrayToIntegrate, &t, (isFirstCall ? 0 : tStepMax), &stepsizeSuggested, smallestAllowableStepsize, absError, relError );
      isIntegrationFinished = errorMessage  ||  (isIntegrateForward && t > tFinalMinusEpsilon)  ||  (!isIntegrateForward && t < tFinalMinusEpsilon);
      if( (isPrintToScreen = printIntScreen  &&  (isIntegrationFinished || --printCounterScreen <= 0))  != 0 ) printCounterScreen = printIntScreen;
      if( (isPrintToFile   = printIntFile    &&  (isIntegrationFinished || --printCounterFile   <= 0))  != 0 ) printCounterFile   = printIntFile;
      OutputToScreenOrFile( t, outputToFill, isPrintToScreen, isPrintToFile );
      isFirstCall = 0;
   }
   return errorMessage;
}

File: MGToolbox/MGProgramFiles/MGIntegrator.f
C********************************************************************************
C** PURPOSE:  Solve a set of first order ordinary differential equations of    **
C**           the form dy(i)/dt = F(t,y(1), ..., y(numY) (i = 1, ..., numY).   **
C**                                                                            **
C** INPUT:                                                                     **
C**     eqns: Subroutine that evaluates dy(i)/dt (i = 1, ..., numY), the       **
C**           first derivatives of y(1), ..., y(numY) with respect to t.       **
C**                                                                            **
C**     numY: The number of differential equations to be solved.               **
C**                                                                            **
C**        y: One-dimensional array whose elements are y(1), ..., y(numY).     **
C**                                                                            **
C**        t: Independent variable.                                            **
C**                                                                            **
C**    tStep: Maximum integration stepsize.                                    **
C**                                                                            **
C** absError: Allowable absolute error in y(i)  (i=1, ..., numY).              **
C**                                                                            **
C** relError: Allowable relative error in y(i)  (i=1, ..., numY).              **
C**                                                                            **
C**      com: When com = 2, the Kutta-Merson algorithm (L. Fox, Numerical      **
C**           Solutions of Ordinary and Partial Differential Equations,        **
C**           Palo Alto: Addison-Wesley, 1962, pp. 24-25) is employed to       **
C**           perform the numerical solution of the differential equations.    **
C**           Accordingly, dy(i)/dt (i = 1, ..., numY) are evaluated at        **
C**           every integration boundary, including those at tInitial,         **
C**           tFinal, and ones created when tStep is halved to satisfy the     **
C**           requirements imposed by absError and relError.  Integration      **
C**           is self-starting at each boundary, and the occurrence, at        **
C**           boundaries, of discontinuities in derivatives does not lead      **
C**           to failure of the integration process.                           **
C**                                                                            **
C**           When com = 1, a modified form of the Kutta-Merson algorithm      **
C**           is employed.  It is nearly 20% faster than the one used when     **
C**           com = 2 because no recalculation of derivatives at inte-         **
C**           gration boundaries between tInitial and tFinal takes place.      **
C**           Integration is self-starting at tInitial and tFinal only.        **
C**           Integration may fail if any of dy(i)/dt (i = 1, ..., numY)       **
C**           is discontinuous between tInitial and tFinal.                    **
C**                                                                            **
C**           When com = 0, the function eqns is called and dy(i)/dt           **
C**           (i = 1, ..., numY) are evaluated, but no integration             **
C**           is performed.                                                    **
C**                                                                            **
C** OUTPUT:   The value of  t + tStep  is returned in t, and the values of     **
C**           y(i) at  t + tStep  are returned in y.                           **
C**                                                                            **
C** Portions copyright (c) 1988-2021 by Paul Mitiguy.  Use permitted under the **
C**     3-clause BSD license: https://opensource.org/licenses/BSD-3-Clause.    **
C**     This copyright notice must appear in all copies and distributions.     **
C**     This particular copyright notice applies only to the KUTTA subroutine. **
C********************************************************************************
      SUBROUTINE KUTTA (eqns,numY,Y,T,tStep,absError,relError,com,*)
      EXTERNAL         eqns
      INTEGER          numY, com, NUMCUTS, I
      LOGICAL          STEPDBL, ENTRY
      DOUBLE PRECISION Y(numY), F0, F1, F2, Y1, Y2
      DOUBLE PRECISION T, tStep, absError, relError, ERROR, TEST
      DOUBLE PRECISION tFinal, TT, HC, H, H2, H3, H6, H8
      COMMON/CKUTTA/   F0(100),F1(100),F2(100),Y1(100),Y2(100)
      DATA             HC, NUMCUTS / 0.0D0, 20 /

C**   Make sure F0, F1, F2, Y1, Y2 arrays are sufficiently dimensioned
      if( numY .GE. 100 ) THEN
        WRITE(*,2020)
        RETURN 1
      ENDIF

C**   If com=0, call eqns subroutine and return.
      IF( com .EQ. 0) THEN
        CALL eqns(T, Y, F0, 1)
        RETURN
      ENDIF

C**   Check for initial entry and adjust current value of stepsize.
      IF(numY .EQ. 0) THEN
        HC = tStep
        RETURN
      ENDIF
      IF(tStep .EQ. 0) RETURN 1
      IF(HC*tStep .LT. 0) HC = -HC
      IF(HC .EQ. 0)       HC = tStep

C**   Set local variables
      H = HC
      TT = T + H
      tFinal = T + tStep
      T  = tFinal
      ENTRY = .TRUE.

C**   Check round-off problems.
100   IF( TT+H .EQ. TT ) THEN
        T = TT
        WRITE(*,2010) H, T
        CALL eqns(T, Y, F0, 0)
        RETURN 1
      ENDIF
C**   Main Kutta-Merson step
      H2 = H * 0.5D0
      H3 = H / 3.0D0
      H6 = H / 6.0D0
      H8 = H * 0.125D0
      IF( com .EQ. 2 .OR. ENTRY )  CALL eqns(TT-H, Y, F0, 1)
      ENTRY = .FALSE.
      DO 110  I=1,numY
110     Y1(I) = Y(I) + H3*F0(I)
      CALL eqns(TT-2.0*H3, Y1, F1, 0)
      DO 120  I=1,numY
120     Y1(I) = Y(I) + H6*(F0(I) + F1(I))
      CALL eqns(TT-2.0*H3, Y1, F1, 0)
      DO 130  I=1,numY
130     Y1(I) = Y(I) + H8*(F0(I) + 3.0D0*F1(I) )
      CALL eqns(TT-H2,     Y1, F2, 0)
      DO 140  I=1,numY
140     Y1(I) = Y(I) + H2*(F0(I) - 3.0D0*F1(I)+ 4.0D0*F2(I) )
      CALL eqns(TT,        Y1, F1, 0)
      DO 150  I=1,numY
150     Y2(I) = Y(I) + H6*(F0(I) +  4.0D0*F2(I) + F1(I) )
C**   Assume that step needs to be doubled.  Check error criterion
      STEPDBL = .TRUE.
      DO 160 I=1,numY
        ERROR = DABS(Y1(I) - Y2(I)) * 0.2D0
        TEST  = DABS(Y1(I)) * relError
        IF(ERROR .GE. TEST .AND. ERROR .GE. absError) THEN
          HC = H2
          H  = HC
          TT = TT - H2
          NUMCUTS = NUMCUTS - 1
          IF(NUMCUTS .GE. 0) GO TO 100
          T = TT - H
          WRITE(*,2000) T
          CALL eqns(T, Y, F0, 0)
          RETURN 1
        ENDIF
      IF(STEPDBL .AND. 64.0D0*ERROR .GT. TEST
     &           .AND. 64.0D0*ERROR .GT. absError) STEPDBL=.FALSE.
160   CONTINUE
      DO 170  I = 1,numY
170     Y(I) = Y2(I)
C**   Double the STEPSIZE, maybe.
      IF( STEPDBL .AND. DABS(H+H) .LE. DABS(tStep) .AND.
     &     DABS(TT+H+H) .LE. DABS(tFinal) )  THEN
        HC = H + H
        H  = HC
        NUMCUTS = NUMCUTS + 1
      ENDIF
      IF( TT .EQ. tFinal ) THEN
        CALL eqns(tFinal, Y, F0, 2)
        RETURN
      ENDIF
      TT = TT + H
      IF( (H .GT. 0 .AND. TT .GT. tFinal-0.1D0*H) .OR.
     &    (H .LT. 0 .AND. TT .LT. tFinal-0.1D0*H)  )  THEN
        H  = tFinal - (TT-H)
        TT = tFinal
      ENDIF
      IF( com .EQ. 1 ) THEN
        DO 180  I = 1,numY
180       F0(I) = F1(I)
      ENDIF
      GOTO 100

2000  FORMAT(/1X,'THE STEPSIZE HAS BEEN HALVED TOO MANY TIMES; T = ',
     &1PD12.4,/1X,'ERROR: NUMERICAL INTEGRATION FAILED TO CONVERGE.',//)
2010  FORMAT(/1X,'THE STEPSIZE OF ',1PD22.14,' IS TOO SMALL RELATIVE ',
     &'TO THE TERMINAL TIME OF',/1PD22.14,'.  INTEGRATION HALTED BECA',
     &'USE OF NUMERICAL ROUND-OFF.',/,'THE STEPSIZE MAY HAVE BEEN CUT ',
     &'TOO MANY TIMES.'//)
2020  FORMAT(/1X,'ERROR: INCREASE THE ARRAY SIZE IN KUTTA.'//)
      END

